File: src/workers/methods/enkf.js
import { matInverse, periodicDist, gaspariCohn, randomNormal } from '../math';

export function updateEnKF(state, y, H, nobs, N, obsIndices, R_diag, R, localization, applyH) {
  const M = state.ensembleSize;
  const ens = state.ensemble;
  const x_mean = state.x_mean;

  // Build perturbation matrix X (N x M): X[:,j] = ens[j] - x_mean
  // Then HX (nobs x M) = rows of X at obsIndices
  const X_at_obs = new Array(nobs); // nobs x M (HX)
  const X_full = new Array(N);      // N x M
  for (let i = 0; i < N; i++) {
    const row = new Float64Array(M);
    for (let j = 0; j < M; j++) row[j] = ens[j][i] - x_mean[i];
    X_full[i] = row;
  }
  for (let i = 0; i < nobs; i++) {
    X_at_obs[i] = X_full[obsIndices[i]];
  }

  // P_e = X * (HX)^T / (M-1) -> N x nobs
  // P_loc[i][j] = P_e[i][j] * GC(dist(i, obsIndices[j]))
  const invM1 = 1 / (M - 1);
  const P_loc = new Array(N);
  for (let i = 0; i < N; i++) {
    const row = new Float64Array(nobs);
    const Xi = X_full[i];
    for (let j = 0; j < nobs; j++) {
      let sum = 0;
      const Xj = X_at_obs[j];
      for (let k = 0; k < M; k++) sum += Xi[k] * Xj[k];
      row[j] = sum * invM1 * gaspariCohn(periodicDist(i, obsIndices[j], N), localization);
    }
    P_loc[i] = row;
  }

  // HPH = HX * (HX)^T / (M-1), then localize
  const S = new Array(nobs);
  for (let i = 0; i < nobs; i++) {
    S[i] = new Float64Array(nobs);
    const Xi = X_at_obs[i];
    for (let j = 0; j < nobs; j++) {
      let sum = 0;
      const Xj = X_at_obs[j];
      for (let k = 0; k < M; k++) sum += Xi[k] * Xj[k];
      S[i][j] = sum * invM1 * gaspariCohn(periodicDist(obsIndices[i], obsIndices[j], N), localization);
    }
    S[i][i] += R_diag + 1e-6; // Add R diagonal + regularization
  }

  const S_inv = matInverse(S);
  
  // K = P_loc * S_inv: N x nobs
  const K = new Array(N);
  for (let i = 0; i < N; i++) {
    const row = new Float64Array(nobs);
    const Pi = P_loc[i];
    for (let j = 0; j < nobs; j++) {
      let sum = 0;
      for (let k = 0; k < nobs; k++) sum += Pi[k] * S_inv[k][j];
      row[j] = sum;
    }
    K[i] = row;
  }

  // Update each ensemble member
  for (let i = 0; i < M; i++) {
    const y_pert = new Array(nobs);
    for (let j = 0; j < nobs; j++) y_pert[j] = y[j] + randomNormal() * Math.sqrt(R_diag);
    const Hx = applyH(ens[i]);
    const innov = new Array(nobs);
    for (let j = 0; j < nobs; j++) innov[j] = y_pert[j] - Hx[j];
    for (let j = 0; j < N; j++) {
      let sum = 0;
      for (let k = 0; k < nobs; k++) sum += K[j][k] * innov[k];
      ens[i][j] += sum;
    }
  }
}

File: crates/eduda_wasm/src/methods/enkf.rs
use crate::math::{mat_inverse, periodic_dist, gaspari_cohn, Rng};

pub fn update_enkf(
    ensemble: &mut [f64], // M * N flat slice (ensemble[m * N + i])
    x_mean: &[f64],
    y: &[f64],
    obs_indices: &[usize],
    r_diag: f64,
    localization: f64,
    m: usize,
    n: usize,
    rng: &mut Rng,
) {
    let nobs = obs_indices.len();
    if nobs == 0 { return; }

    let inv_m1 = 1.0 / (m - 1) as f64;
    let r_std = r_diag.sqrt();

    // Perturbation matrix X (N x M): X[i, j] = ensemble[j * N + i] - x_mean[i]
    let mut x_pert = vec![0.0; n * m];
    for i in 0..n {
        for j in 0..m {
            x_pert[i * m + j] = ensemble[j * n + i] - x_mean[i];
        }
    }

    // P_loc: N x nobs
    let mut p_loc = vec![0.0; n * nobs];
    for i in 0..n {
        for j in 0..nobs {
            let oj = obs_indices[j];
            let mut sum = 0.0;
            for k in 0..m {
                sum += x_pert[i * m + k] * x_pert[oj * m + k];
            }
            let gloc = gaspari_cohn(periodic_dist(i, oj, n), localization);
            p_loc[i * nobs + j] = sum * inv_m1 * gloc;
        }
    }

    // S: nobs x nobs
    let mut s = vec![0.0; nobs * nobs];
    for i in 0..nobs {
        let oi = obs_indices[i];
        for j in 0..nobs {
            let oj = obs_indices[j];
            let mut sum = 0.0;
            for k in 0..m {
                sum += x_pert[oi * m + k] * x_pert[oj * m + k];
            }
            let gloc = gaspari_cohn(periodic_dist(oi, oj, n), localization);
            s[i * nobs + j] = sum * inv_m1 * gloc;
        }
        s[i * nobs + i] += r_diag + 1e-6;
    }

    let mut s_inv = vec![0.0; nobs * nobs];
    mat_inverse(&s, nobs, &mut s_inv);

    // K = P_loc * S_inv: N x nobs
    let mut k_mat = vec![0.0; n * nobs];
    for i in 0..n {
        for j in 0..nobs {
            let mut sum = 0.0;
            for k in 0..nobs {
                sum += p_loc[i * nobs + k] * s_inv[k * nobs + j];
            }
            k_mat[i * nobs + j] = sum;
        }
    }

    // Update each ensemble member with perturbed observations
    let mut innov = vec![0.0; nobs];
    for j in 0..m {
        for k in 0..nobs {
            let y_pert = y[k] + rng.normal() * r_std;
            let hx = ensemble[j * n + obs_indices[k]];
            innov[k] = y_pert - hx;
        }

        for i in 0..n {
            let mut sum = 0.0;
            for k in 0..nobs {
                sum += k_mat[i * nobs + k] * innov[k];
            }
            ensemble[j * n + i] += sum;
        }
    }
}
